### PR TITLE
Protect prompt buffer against all conditions

### LINKED ISSUE #1976: Unhandled USOCKET:TIMEOUT-ERROR in thread
**Describe the bug**

1. Install nyxt from `nyxt-browser-git` AUR package
2. start nyxt
3. Open some random pages
4. Crashes 

**Information**
- OS name+version: `Linux arch 5.15.11-arch2-1 #1 SMP PREEMPT Wed, 22 Dec 2021 09:23:54 +0000 x86_64 GNU/Linux`
- Graphics card and driver:`AMD ATI 03:00.0 Renoir`
- Desktop environment / Window manager name+version:`sway version 1.6.1`
- How you installed Nyxt (Guix pack, package manager, build from source): AUR package, `nyxt-browser-git`
- Information from `show-system-information`:

```
Nyxt version: 2.1.1-1192-g5cd7dbdd
Renderer version: GI-GTK
Operating system kernel: Linux 5.15.11-arch2-1
Lisp implementation: SBCL 2.1.11 (Dynamic space size: 1073741824)
Features: (:WEBKIT2 :WEBKIT2-2.34 :WEBKIT2-PASTE-PLAINTEXT :WEBKIT2-TRACKING
 :WEBKIT2-MUTE :WEBKIT2-EMOJI :WEBKIT2-MEDIA :WEBKIT2-SANDBOXING :GTK-3-22
 :GTK-3-20 :GTK-3-18 :GTK-3-16 :GTK-3-14 :GTK-3-12 :GTK-3-10 :GTK-3-8 :GTK-3-6
 :GTK-3-4 :GTK :GDK-3-22 :GDK-3-20 :GDK-3-18 :GDK-3-16 :GDK-3-14 :GDK-3-12
 :GDK-3-10 :GDK-3-8 :GDK-3-6 :GDK-3-4 :CAIRO-1-10 :CAIRO-1-12 :GDK-PIXBUF
 :GLIB-2-30 :GLIB-2-32 :GLIB-2-34 :GLIB-2-36 :GLIB-2-38 :GLIB-2-40 :GLIB-2-42
 :GLIB-2-44 :GLIB-2-46 :GLIB-2-48 :GLIB-2-50 :GLIB-2-52 :GLIB-2-54 :GLIB-2-56
 :GLIB-2-58 :GLIB :NYXT-2 :FSET-EXT-STRINGS :CUSTOM-HASH-TABLE-NATIVE :SWANK
 :PLUMP-UTF-32 :GLOBAL-VARS :DECLARE-TYPES :PARENSCRIPT
 :SBCL+SAFE-STANDARD-READTABLE :NAMED-READTABLES :LPARALLEL :21BIT-CHARS
 :CHUNGA :CLOSER-MOP :CL-PPCRE-UNICODE :FLEXI-STREAMS :CL-UNICODE :CL-PPCRE
 :CL-JSON-DOUBLE-FLOAT-IS-SUBSUMED :CL-JSON-SINGLE-FLOAT-IS-SUBSUMED
 :BORDEAUX-THREADS :LPARALLEL.WITH-CLTL2 :LPARALLEL.WITH-CAS
 :LPARALLEL.WITH-STEALING-SCHEDULER :SPLIT-SEQUENCE CHIPZ-SYSTEM:GRAY-STREAMS
 CFFI-FEATURES:FLAT-NAMESPACE CFFI-FEATURES:X86-64 CFFI-FEATURES:UNIX :CFFI
 CFFI-SYS::FLAT-NAMESPACE ALEXANDRIA::SEQUENCE-EMPTYP :FAST-IO-SV :FAST-IO
 :SBCL-USES-SB-ROTATE-BYTE :CL-JSON-CLOS :CL-JSON :THREAD-SUPPORT :ASDF3.3
 :ASDF3.2 :ASDF3.1 :ASDF3 :ASDF2 :ASDF :OS-UNIX :NON-BASE-CHARS-EXIST-P
 :ASDF-UNICODE :X86-64 :GENCGC :64-BIT :ANSI-CL :COMMON-LISP :ELF
 :IEEE-FLOATING-POINT :LINUX :LITTLE-ENDIAN :PACKAGE-LOCAL-NICKNAMES
 :SB-CORE-COMPRESSION :SB-LDB :SB-PACKAGE-LOCKS :SB-THREAD :SB-UNICODE :SBCL
 :UNIX)

ASDF version: 3.3.1
ASDF registries: (NYXT-SOURCE-REGISTRY ENVIRONMENT-SOURCE-REGISTRY)
Critical dependencies: (/home/lyh/.cache/paru/clone/nyxt-browser-git/src/nyxt-browser/_build/cl-cffi-gtk/gtk/cl-cffi-gtk.asd
 /home/lyh/.cache/paru/clone/nyxt-browser-git/src/nyxt-browser/_build/cl-gobject-introspection/cl-gobject-introspection.asd
 /home/lyh/.cache/paru/clone/nyxt-browser-git/src/nyxt-browser/_build/cl-webkit/webkit2/cl-webkit2.asd)
```

**Output when started from a shell**
```
<INFO> [21:40:16] Listening to socket "/run/user/1000/nyxt/nyxt.socket".
Nyxt version 2.1.1-1192-g5cd7dbdd
<INFO> [21:40:19] Loading history of 7 URLs from "/home/lyh/.local/share/nyxt/history/default.lisp".
<INFO> [21:40:21] Loading "https://duckduckgo.com/".
Unhandled USOCKET:TIMEOUT-ERROR in thread #<SB-THREAD:THREAD "Prompter update thread" RUNNING
                                             {1006D37443}>:
  Condition USOCKET:TIMEOUT-ERROR was signalled.

Backtrace for: #<SB-THREAD:THREAD "Prompter update thread" RUNNING {1006D37443}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<USOCKET:TIMEOUT-ERROR {10074EB153}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK SB-EXT:*INVOKE-DEBUGGER-HOOK* #<USOCKET:TIMEOUT-ERROR {10074EB153}>)
2: (INVOKE-DEBUGGER #<USOCKET:TIMEOUT-ERROR {10074EB153}>)
3: (ERROR USOCKET:TIMEOUT-ERROR :SOCKET #<USOCKET:STREAM-USOCKET {10074DEBC3}>)
4: (USOCKET::HANDLE-CONDITION #<SB-EXT:TIMEOUT {10074E3FA3}> #<USOCKET:STREAM-USOCKET {10074DEBC3}> "duckduckgo.com")
5: (SB-KERNEL::%SIGNAL #<SB-EXT:TIMEOUT {10074E3FA3}>)
6: (ERROR SB-EXT:TIMEOUT)
7: (USOCKET:SOCKET-CONNECT "duckduckgo.com" 443 :PROTOCOL :STREAM :ELEMENT-TYPE (UNSIGNED-BYTE 8) :TIMEOUT 10 :DEADLINE NIL :NODELAY T :LOCAL-HOST NIL :LOCAL-PORT NIL)
8: ((LABELS DEXADOR.BACKEND.USOCKET::MAKE-NEW-CONNECTION :IN DEXADOR.BACKEND.USOCKET:REQUEST) #<QURI.URI.HTTP:URI-HTTPS https://#1=duckduckgo.com/ac/?q=https%3A%2F%2Fwiki.gnome.org%2FApps%2FWeb>)
9: (DEXADOR.BACKEND.USOCKET:REQUEST #<unavailable argument> :METHOD :GET)
10: ((LAMBDA (URL &REST NYXT::ARGS) :IN "/home/lyh/.cache/paru/clone/nyxt-browser-git/src/nyxt-browser/source/search-engine.lisp") "https://duckduckgo.com/ac/?q=https%3A%2F%2Fwiki.gnome.org%2FApps%2FWeb")
11: ((LAMBDA (NYXT::INPUT) :IN MAKE-SEARCH-COMPLETION-FUNCTION) "https://wiki.gnome.org/Apps/Web")
12: (NYXT::INPUT->QUERIES "https://wiki.gnome.org/Apps/Web" :CHECK-DNS-P T :ENGINE-COMPLETION-P T)
13: ((LAMBDA NIL :IN PROMPTER::UPDATE))
14: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
15: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
16: ((FLET "WITHOUT-INTERRUPTS-BODY-10" :IN SB-THREAD::RUN))
17: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
18: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-THREAD::RUN))
19: (SB-THREAD::RUN)
20: ("foreign function: call_into_lisp")
21: ("foreign function: funcall1")

unhandled condition in --disable-debugger mode, quitting
<WARN> [21:40:31] Warning: Error: Joining thread failed: thread #&lt;THREAD &quot;Nyxt renderer thread&quot; ABORTED {10081E2A53}&gt; did not return normally.
```

### PR DESCRIPTION
**Pull request by Ambrevar**
This avoids crashes when the caller runs a prompter with callbacks that raise
unhandled conditions.

Fixes #1976.

### CODE REVIEW

**Comment by Ambrevar** [comment]
This backports the `run-thread` macro from Nyxt to the prompter library.

However we can't use `*run-from-repl-p*` here, so all conditions are systematically caught and outputted as warnings.

Is this the right behaviour?  Shall we add a configuration knob to make it easy to choose what to do on error?
This way, Nyxt could configure the library to behave differently when `*run-from-repl-p*`.  Thoughts?  Over-engineered?

**Comment by Ambrevar** [comment]
To test this, you can set the following slot of the `new-url-or-search-source`:

```lisp
(prompter:filter-postprocessor
    (lambda (suggestions source input)
      (declare (ignore suggestions source))
      (/ 0 input)))
```

**Comment by aartaka** [comment]
> Is this the right behaviour? Shall we add a configuration knob to make it easy to choose what to do on error? This way, Nyxt could configure the library to behave differently when `*run-from-repl-p*`. Thoughts? Over-engineered?

Some Lisp libraries I've seen have a special variable (usually something like `lib-name:*debug-ob-error*`) to manifest whether to raise the condition to debugger or supress it. I think this convention can work in case of `prompter`.

**Comment by Ambrevar** [comment]
There is also `*invoke-debugger-hook*`, any idea if we could leverage it?

**Comment by Ambrevar** [comment]
I've added `*debug-on-error*` for now.

Do you know of any library that uses `*debug-on-error*`?

**Comment by aartaka** [comment]
> I've added `*debug-on-error*` for now.
> 
> Do you know of any library that uses `*debug-on-error*`?

Mostly testing frameworks, like Prove or 5AM, as they can benefit from the debugger. Our case of development setup is quite close to those.

**Comment by Ambrevar** [comment]
This patch set now also contains important fixes for the prompter.

File: libraries/prompter/prompter-source.lisp [hunk 1]
```diff
@@ -492,31 +492,65 @@ If you are looking for a source that just returns its plain suggestions, use `so
...
-       (setf (slot-value source 'suggestions) (initial-suggestions source))
-       (bt:release-lock (initial-suggestions-lock source))
-       (when (listp (constructor source))
-         ;; Initial suggestions are set synchronously in this case.
-         (calispel:! wait-channel t)))
-     :name "Prompter source init thread")
+    (run-thread "Prompter source init thread"
+      (bt:acquire-lock (initial-suggestions-lock source))
+      ;; `initial-suggestions' initialization must be done before first input can be processed.
+      (cond ((listp (constructor source))
```
Comment (aartaka) [inline libraries/prompter/prompter-source.lisp:538]: `typecase`? `etypecase`? `constructor` can only be list or function, after all.

File: libraries/prompter/prompter-source.lisp [hunk 2]
```diff
@@ -492,31 +492,65 @@ If you are looking for a source that just returns its plain suggestions, use `so
...
+         (declare (ignorable ,c))
+         ,(let* ((condition-index (position :condition args))
+                 (new-args (if condition-index
+                               (append (subseq args 0 condition-index)
+                                       `(,c)
+                                       (subseq args (1+ condition-index)))
+                               args)))
+            `(warn ,format-string ,@new-args))))))
+
+(defmacro run-thread (&body body)   ; TODO: Copied from Nyxt.  Move to serapeum?
```
Comment (aartaka) [inline libraries/prompter/prompter-source.lisp:513]: Now that we're doing the last 2-series release, we can finally have the `name` positional argument to `run-thread` and stop parsing the body looking for the name of the thread.

Comment (Ambrevar) [reply libraries/prompter/prompter-source.lisp:513]: You mean for 3.0?  Yes we can! :)

**Pull request merged**